File: cpp/src/binning.cpp
using namespace std;

#include <iostream>
#include <numeric>
#include <cmath>
#include <cstring> 
#include <vector>
#include <bitset>
#include <sstream>
#include <assert.h>
#include <algorithm>


#include "binning.h"


void Binning::get_logarithmic_bins(int start, int divisor, int n, int* bins)
{
  float bin_width = start;
  for(int i = n-1; i >= 0; i--)
  {
      bins[i] = int(round(bin_width));
      bin_width /= divisor;
  }
}


void Binning::get_equal_bins(int width, int n, int* bins)
{
    std::fill(bins, bins+n, width);
}


void Binning::get_linear_bins(int start, int stop, int step, int* bins)
{
    int i = 0;
    int bin_with = start;
    while(bin_with < stop)
    {
        bins[i] = bin_with;
        bin_with += step;
        i++;
    }
}


long Binning::compute_binning(double* Sy, int len_Sy, double* Sx, int len_Sx, int* bins, int len_bins, int T)
{
    k = len_bins;

    prob_map = new double[(1 << (k + k)) * sizeof(double)];
    memset(prob_map, 0, (1 << (k + k)) * sizeof(double)); 
    
    sort(Sy, Sy + len_Sy);
    sort(Sx, Sx + len_Sx);


    double* S_total = new double[(len_Sy+len_Sx)*sizeof(double)];
    memcpy(S_total, Sy, len_Sy*sizeof(double));
    memcpy(S_total+len_Sy, Sx, len_Sx*sizeof(double));
    sort(S_total, S_total + len_Sy + len_Sx);


    int idxY = 0;
    int idxX = 0;
    long w = accumulate(bins, bins + k, 0); 

    for(int idx = 0; idx < (len_Sy+len_Sx); idx++)
    {
        double val = S_total[idx];
        for(int i = (val-w+1); i < (val+1); i++)
        {
            if((i+w) == T+2) 
                break;
            
            if(i < 0) 
                continue;
            
            if(idx > 0 && ((i <= S_total[idx-1]) && (S_total[idx-1] < (i+w))))
                continue;
            
            int comb = 0;

            int idxY_tmp = idxY;
            while((idxY_tmp < len_Sy) && (Sy[idxY_tmp] < (i+w)))
            {
                int sumbins = 0;
                for(int bin_idx = len_bins-1; bin_idx > -1; bin_idx--)
                {
                    if(((i+sumbins) <= Sy[idxY_tmp]) && (Sy[idxY_tmp] < (i+sumbins+bins[bin_idx])))
                       comb |= (1 << (2 * len_bins - bin_idx - 1)); 
                    sumbins += bins[bin_idx];
                }
                idxY_tmp++;
            }

            int idxX_tmp = idxX;
            while((idxX_tmp < len_Sx) && (Sx[idxX_tmp] < (i+w)))
            {
                int sumbins = 0;
                for(int bin_idx = len_bins-1; bin_idx > -1; bin_idx--)
                {
                    if(((i+sumbins) <= Sx[idxX_tmp]) && (Sx[idxX_tmp] < (i+sumbins+bins[bin_idx])))
                        comb |= (1 << (len_bins - bin_idx - 1));
                    sumbins += bins[bin_idx];
                }
                idxX_tmp++;
            } 
            prob_map[comb]++;
        }

        if((idxY < len_Sy) && S_total[idx] == Sy[idxY])
            idxY++;
        else
            idxX++;
    }

    int p_sum =  accumulate(prob_map, prob_map + (1 << (k + k)), 0);
    prob_map[0] = T - w - p_sum + 2;
    for(int i = 0; i < (1 << (k + k)); i++)
        prob_map[i] /= (T-w+2);
    
    delete [] S_total;

    return T - w + 2;
}


void Binning::eliminate_bin(int eliminate_bins, double* reduced_prob_map)
{   
    for(int prob_idx = 0; prob_idx < (1 << (k + k)); prob_idx++) //iterate over probability map
    {
        int idx = 0;
        for(int bin_idx = 0; bin_idx < k + k; bin_idx++) //iterate over all bins
        {
            if((eliminate_bins >> (k + k - 1 - bin_idx) & 1))
            {
                int b = prob_idx >> (k+k-(bin_idx+1)) & 1;
                idx = idx << 1;
                idx |= b;
            }
        }
        reduced_prob_map[idx] += prob_map[prob_idx];
    }
}

void Binning::eliminate_bin(int eliminate_bins, double* reduced_prob_map, double* prob_map)
{   
    for(int prob_idx = 0; prob_idx < (1 << (k + k)); prob_idx++) //iterate over probability map
    {
        int idx = 0;
        for(int bin_idx = 0; bin_idx < k + k; bin_idx++) //iterate over all bins
        {
            if((eliminate_bins >> (k + k - 1 - bin_idx) & 1))
            {
                int b = prob_idx >> (k+k-(bin_idx+1)) & 1;
                idx = idx << 1;
                idx |= b;
            }
        }
        reduced_prob_map[idx] += prob_map[prob_idx];
    }
}


Binning::~Binning()
{
    delete[] prob_map;
}

File: py/binning.py
"""Module binning
   Converts a list of timestamps into a binned RV
"""



import numpy as np
import itertools
import time
from decimal import Decimal, ROUND_HALF_UP





def get_logarithmic_bins(start, divisor, n):
    """Returns a list of logarithmicaly distributed bin widths
       Example: start=20, divisor=2, n=3 returns [5,10,20]
    
    Arguments:
        start {int} -- width of largest bin
        divisor {int} -- bin width decrement
        n {int} -- number of bins

    Returns:
        list of int -- bin widths
    """
    bin_width = start
    bins = []
    for i in range(0, n):
        bins.append(int(Decimal(bin_width).to_integral_value(rounding=ROUND_HALF_UP)))
        bin_width = bin_width / divisor
    bins.reverse()
    return bins
    

def get_equal_bins(width, n):
    """Returns a list of equally sized bins with a constant width
       Example: width=4, n=3 returns [4,4,4,4]
    
    Arguments:
        width {int} -- width of bin
        n {int} -- number of bins to generate
    
    Returns:
        list of int -- bin widths
    """
    return [width] * n


def get_linear_bins(start, stop, step):
    """Returns a list bins with a change in size step.
    The smallest bin has size start, the largest bin has size stop.
    
    Arguments:
        start {int} -- size of largest bin
        stop {int} -- size of smallest bin
        step {int} -- size difference between each two bins
    
    Returns:
        list of int -- bin widths
    """
    return np.arange(start=start, stop=stop, step=step).tolist()


def compute_binning(Sy, Sx, bins, T):
    """Computes probability P(Yt, Yt-1, ... Yt-k, Xt-1, ... Xt-l)
    
    Arguments:
        Sy {list of int} -- timestamp list of user Y, S[0] = timestamp at time t0, S[1] = timestamp at time t1, ...
        Sx {list of int} -- timestamp list of user X, S[0] = timestamp at time t0, S[1] = timestamp at time t1, ...
        bins {list of int} -- bin widths in seconds, bin[0] = bin width at time t0, bin[1] = bin width at time t-1, bin[2] = bin width at time t-2, ...
        T {int} -- upper time boundary
    
    Returns:
        dict of tuple: float -- probability map, key: (Yt, Yt-1, Yt-2, ... Yt-k, Xt-1, Xt-2, ... X-l)
    """
    w = sum(bins)
    p = dict.fromkeys(itertools.product([1,0], repeat=len(bins)*2), 0)
    Sy = sorted(Sy)
    Sx = sorted(Sx)
    S_total = sorted(Sy + Sx)

    idxY = idxX = 0

    for idx in range(0, len(S_total)):
        val = S_total[idx]
       

        for i in range(val-w+1, val+1):

            if i+w == T+2:
                break
            
            if i < 0:
                continue
            

            if idx > 0 and i <= S_total[idx-1] < i+w:
                
                continue

            comb = [0] * (len(bins) * 2)

            idxY_tmp = idxY
            while idxY_tmp < len(Sy) and Sy[idxY_tmp] < (i+w):
                sumbins = 0
                for bin_idx in range(len(bins)-1, -1, -1):
                    if(i + sumbins) <= Sy[idxY_tmp] < (i + sumbins + bins[bin_idx]):
                        comb[bin_idx] = 1
                    sumbins += bins[bin_idx]
                idxY_tmp +=1
            
            idxX_tmp = idxX
            while idxX_tmp < len(Sx) and Sx[idxX_tmp] < (i+w):
                sumbins = 0
                for bin_idx in range(len(bins)-1, -1, -1):
                    if(i+ sumbins) <= Sx[idxX_tmp] < (i+sumbins+bins[bin_idx]):
                        comb[bin_idx+len(bins)] = 1
                    sumbins += bins[bin_idx]
                idxX_tmp +=1

            p [tuple(comb)] += 1

        if idxY < len(Sy) and S_total[idx] == Sy[idxY]:
            idxY+=1
        elif idxX < len(Sx) and S_total[idx] == Sx[idxX]:
            idxX+=1
        else:
            assert(False)
    p_sum = sum(p.values())
    p[tuple([0]*len(bins)*2)] = T - w - p_sum + 2
    total = T - w + 2
    return {k: v / total for k, v in p.items()}, total                

def eliminate_bin(p, eliminate_bins):
    """Marginalizes probability distribution by eliminating bins with indices in eliminate_bins

    Arguments:
        p {dictionary of tuple: float} -- probability map
        eliminate_bins {list of int} -- indices of bins to be eliminated, for exmample: bins = [1,2,3], eliminate_bins = [0] gives new bins = [2,3]
    
    Returns:
        p_eliminated {dictionary of tuple: float} -- probability map
    """
    p_eliminated = {}
    for comb in p:
        comb_eliminated = tuple([ comb[i] for i in range(0, len(comb)) if i not in eliminate_bins])
        
        p_eliminated[comb_eliminated] = p[comb] if comb_eliminated not in p_eliminated else p[comb] +  p_eliminated[comb_eliminated]
    return p_eliminated
